shoebot/data/typography.py

```diff
@@ -178,7 +194,23 @@ def _pre_render(self):
         # self.layout.set_spacing(int(((self.lineheight-1)*self._fontsize)*Pango.SCALE)) #pango requires an int casting
         # we pass pango font description and the text to the pango layout
         self._pango_layout.set_font_description(self._pango_fontface)
-        self._pango_layout.set_text(self.text, -1)
+
+        if not self.markup_vars:
+            self._pango_layout.set_text(self.text, -1)
+        else:
+            # some of the specified settings require a Pango.Markup hack
+            # see https://stackoverflow.com/questions/55533312/how-to-create-a-letter-spacing-attribute-with-pycairo
+            # and https://developer.gnome.org/pango/1.46/pango-Markup.html
+
+            # we want to output something like
+            # <span letter_spacing="2048">Hello World</span>
+            markup = '<span'
+            # markup_vars was set in __init__()
+            for setting, value in self.markup_vars.items():
+                markup += f' {setting}="{self.markup_vars[setting]}"'
+            markup += f">{self.text}</span>"
+            self._pango_layout.set_markup(markup)
+
```

Comment:
I need to find some time to stare at this, it's def the sort of thing where tests are needed (think of how much trouble we got into with the nodebox colour lib)

Reply:
Can you reply with some examples of the expected inputs (parameters to the method) and what the markup they should output ?

We can use this for writing some tests + find if there are any edge cases.

Reply:
This is the format:
```
Inconsolata @wght=200
Inconsolata @wght=200,wdth=100
```

Reply:
Cheers, so it would be:
```
 <span "Inconsolata @wght=200">blah blah</span>
 <span "Inconsolata @wght=200,wdth=100">blah blah</span>  
```

Reply:
ah yes, exactly that! I could have typed out the whole thing, sorry

Reply:
np, unit tests are the definition of pedantry so it's good to capture all the details.

Reply:
wait, according to the [current docs](https://lazka.github.io/pgi-docs/Pango-1.0/), the syntax should be:
```
"Cantarell Italic Light 15 `wght`=200"
```
we should update this (i can do it, just not now)

Reply:
OK so I tested this syntax and it doesn't seem to work. I'll stick to the @-syntax and write a small test to ensure markup is correct

Reply:
I see no way to get the markup in Pango.Layout (only set), and we're not holding the markup data anywhere, so I'm a bit stumped as how to test this.

Reply:
OR, for a minimal test, we can just query font():

```python
font("Inconsolata", var_wdth=100, var_wght=200*
print(font())
```

gives

```
Inconsolata @wdth=100,wght=200
```

Reply:
That sounds like a sensible test.

I already added `test_text_outputs_pango_text_spans` on this branch, which makes an effort, using mocks to see if we pass through what's expected to Pango.

Reply:
done! making tests is a joy now, this framework is just great. Cheers :)